File: server/src/models/Message.ts
import mongoose, { Schema, Document, Types } from 'mongoose';

export type MessageType = 'text' | 'image' | 'audio' | 'system' | 'file';
export type MessageStatus = 'sent' | 'delivered' | 'read';
export type ClientSource = 'desktop' | 'android' | 'api';
export type ReactionEmoji = '👍' | '❤️' | '😂' | '😮' | '😢' | '😡' | '✅' | '⚠️' | '🙏' | '🎉' | '👋' | '😘';

export const ALLOWED_EMOJIS: ReactionEmoji[] = ['👍', '❤️', '😂', '😮', '😢', '😡', '✅', '⚠️', '🙏', '🎉', '👋', '😘'];

export interface IReaction {
  userId: Types.ObjectId;
  emoji: ReactionEmoji;
  createdAt: Date;
}

export interface IMessage extends Document {
  roomId: Types.ObjectId;
  senderId: Types.ObjectId;
  type: MessageType;
  content: string;
  caption?: string;
  fileName?: string;
  fileSize?: number;
  mimeType?: string;
  fileDeleted?: boolean;
  status: MessageStatus;
  readBy: Types.ObjectId[];
  reactions: IReaction[];
  clientSource?: ClientSource;
  createdAt: Date;
  updatedAt: Date;
}

const MessageSchema = new Schema<IMessage>(
  {
    roomId: {
      type: Schema.Types.ObjectId,
      ref: 'Room',
      required: true,
      index: true,
    },
    senderId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    type: {
      type: String,
      enum: ['text', 'image', 'audio', 'system', 'file'],
      default: 'text',
    },
    content: {
      type: String,
      required: true,
    },
    caption: {
      type: String,
      required: false,
    },
    fileName: {
      type: String,
      required: false,
    },
    fileSize: {
      type: Number,
      required: false,
    },
    mimeType: {
      type: String,
      required: false,
    },
    fileDeleted: {
      type: Boolean,
      default: false,
    },
    status: {
      type: String,
      enum: ['sent', 'delivered', 'read'],
      default: 'sent',
    },
    readBy: [{
      type: Schema.Types.ObjectId,
      ref: 'User',
    }],
    reactions: [{
      userId: {
        type: Schema.Types.ObjectId,
        ref: 'User',
        required: true,
      },
      emoji: {
        type: String,
        required: true,
        enum: ALLOWED_EMOJIS,
      },
      createdAt: {
        type: Date,
        default: Date.now,
      },
    }],
    clientSource: {
      type: String,
      enum: ['desktop', 'android', 'api'],
    },
  },
  {
    timestamps: true,
  }
);

MessageSchema.index({ roomId: 1, createdAt: -1 });
MessageSchema.index({ senderId: 1, createdAt: -1 });

export const Message = mongoose.model<IMessage>('Message', MessageSchema);

File: android/app/src/main/java/com/organizer/chat/data/model/Message.kt
package com.organizer.chat.data.model

import com.google.gson.annotations.SerializedName

data class Message(
    @SerializedName("_id", alternate = ["id"])
    val id: String,
    val roomId: String,
    val senderId: MessageSender,
    val type: String = "text",  // "text", "image", "audio", "system", "file"
    val content: String,
    val caption: String? = null,  // Optional caption for image/file messages
    val fileName: String? = null,  // Original filename for file messages
    val fileSize: Long? = null,  // File size in bytes
    val mimeType: String? = null,  // MIME type (application/pdf, etc.)
    val fileDeleted: Boolean = false,  // True if the file was soft-deleted
    val status: String = "sent",  // "sent", "delivered", "read"
    val readBy: List<String> = emptyList(),
    val reactions: List<Reaction> = emptyList(),
    val clientSource: String? = null,  // "desktop", "android", "api"
    val createdAt: String,
    val updatedAt: String? = null
)

data class Reaction(
    val userId: String,
    val emoji: String,
    val createdAt: String
)

data class MessageSender(
    @SerializedName("_id", alternate = ["id"])
    val id: String,
    val username: String,
    val displayName: String,
    val status: String = "available",
    val statusMessage: String? = null,
    val isOnline: Boolean = false
)

data class MessagesResponse(
    val messages: List<Message>
)

data class SendMessageRequest(
    val roomId: String,
    val type: String = "text",
    val content: String,
    val clientSource: String = "android"
)

data class MessageResponse(
    val message: Message
)

data class ReactRequest(
    val emoji: String
)

data class ReactResponse(
    val message: Message,
    val action: String,
    val roomId: String
)
